tachyon/zk/plonk/halo2/argument_data.h

```diff
@@ -0,0 +1,187 @@
+               std::vector<std::vector<Poly>>&& instance_polys_vec)
+      : advice_columns_vec_(std::move(advice_columns_vec)),
+        advice_blinds_vec_(std::move(advice_blinds_vec)),
+        challenges_(std::move(challenges)),
+        instance_columns_vec_(std::move(instance_columns_vec)),
+        instance_polys_vec_(std::move(instance_polys_vec)) {
+    size_t num_circuits = advice_columns_vec_.size();
+    CHECK_EQ(num_circuits, advice_blinds_vec_.size());
+    CHECK_EQ(num_circuits, instance_columns_vec_.size());
+    CHECK_EQ(num_circuits, instance_polys_vec_.size());
+  }
+
+  template <typename Circuit>
+  static ArgumentData Create(
+      ProverBase<PCS>* prover, std::vector<Circuit>& circuits,
+      const ConstraintSystem<F>& constraint_system,
+      std::vector<std::vector<Evals>>&& instance_columns_vec) {
+    // Generate instance polynomial and write it to transcript.
+    std::vector<std::vector<Poly>> instance_polys_vec =
+        GenerateInstancePolys(prover, instance_columns_vec);
+
+    // Append leading zeros to each column of |instance_columns_vec|.
+    size_t num_circuits = circuits.size();
+    size_t n = prover->pcs().N();
+    for (size_t i = 0; i < num_circuits; ++i) {
+      for (Evals& instance_column : instance_columns_vec[i]) {
+        instance_column.evaluations().resize(n);
+      }
+    }
+
+    // Generate advice poly by synthesizing circuit and write it to transcript.
+    Synthesizer<PCS> synthesizer(num_circuits, &constraint_system);
+    synthesizer.GenerateAdviceColumns(prover, circuits, instance_columns_vec);
+
+    return ArgumentData(std::move(synthesizer).TakeAdviceColumnsVec(),
+                        std::move(synthesizer).TakeAdviceBlindsVec(),
+                        std::move(synthesizer).TakeChallenges(),
+                        std::move(instance_columns_vec),
+                        std::move(instance_polys_vec));
+  }
+
+  size_t GetNumCircuits() const { return instance_columns_vec_.size(); }
+
+  absl::Span<const F> GetAdviceBlinds(size_t circuit_idx) const {
+    CHECK_LT(circuit_idx, GetNumCircuits());
+    return absl::MakeConstSpan(advice_blinds_vec_[circuit_idx]);
+  }
+
+  absl::Span<const F> GetChallenges() const {
+    return absl::MakeConstSpan(challenges_);
+  }
+
+  // Generate a vector of advice coefficient-formed polynomials with a vector
+  // of advice evaluation-formed columns. (a.k.a. Batch IFFT)
+  // And for memory optimization, every evaluations of advice will be released
+  // as soon as transforming it to coefficient form.
+  void TransformAdvice(const Domain* domain) {
+    CHECK(!advice_transformed_);
+    advice_polys_vec_ = base::Map(
+        advice_columns_vec_, [domain](std::vector<Evals>& advice_columns) {
+          return base::Map(advice_columns, [domain](Evals& advice_column) {
+            Poly poly = domain->IFFT(advice_column);
+            // Release advice evals for memory optimization.
+            advice_column = Evals::Zero();
```

Comment:
```suggestion
            advice_column.evaluations().clear();
```
I think this has better readability.

Reply:
But this implies that `evaluataions()` is implemented with `std::vector<T>` and I want to avoid this assumption.